src/runContract.js

```diff
@@ -89,36 +89,71 @@ export default async (event, context) => {
 
     // If a returned XDR has already been submitted throw an error, otherwise we open up a looping vulnerability
 
-    if (!event.headers['X-Turrets'])
-      throw 'Mising X-Turrets header'
-
-    const healthySigners = await Promise.map(uniq([
-      process.env.TURRET_ADDRESS,
-      ...event.headers['X-Turrets'].split(',')
-    ]),
-      async (turret) => server
-      .loadAccount(turret)
-      .then((account) => axios
-        .get(`${tssRoute(account)}/contract/${event.pathParameters.hash}`)
-        .then(async ({data}) => data)
-        .catch(() => null)
+    // TODO: Add an alternative validation method similar to the upload contract payment requirement
+
+    const contractBody = {
+      request: JSON.parse(event.body),
+      signers: []
+    }
+
+    if (event.headers['X-Payment']) {
+      const transaction = new Transaction(event.headers['X-Payment'], Networks[process.env.STELLAR_NETWORK])
+      const hash = transaction.hash().toString('hex')
+
+      if (!find(transaction._operations, {
+        type: 'payment',
+        destination: process.env.TURRET_ADDRESS,
+        amount: new BigNumber(process.env.TURRET_RUN_FEE).toFixed(7),
+        asset: Asset.native()
+      })) throw 'Missing or invalid fee payment'
+
+      await server
+      .transactions()
+      .transaction(hash)
+      .call()
+      .catch((err) => {
+        if (
+          err.response
+          && err.response.status === 404
+        ) return
+
+        else if (err.response)
+          throw err
+
+        throw 'Transaction has already been submitted'
+      })
+
+      await server.submitTransaction(transaction)
+    }
+
+    else if (event.headers['X-Turrets']) {
+      contractBody.signers = await Promise.map(uniq([
+        process.env.TURRET_ADDRESS,
+        ...event.headers['X-Turrets'].split(',')
+      ]),
+        async (turret) => server
+        .loadAccount(turret)
+        .then((account) => axios
+          .get(`${tssRoute(account)}/contract/${event.pathParameters.hash}`)
+          .then(async ({data}) => data)
+          .catch(() => null)
+        )
+      ).then((signers) => chain(signers)
+        .compact()
+        .orderBy(['turret', 'signer', 'fee'], 'desc')
+        .value()
       )
-    ).then((signers) => chain(signers)
-      .compact()
-      .orderBy(['turret', 'signer', 'fee'], 'desc')
-      .value()
-    )
+    }
+
+    else throw 'Cannot run without either an X-Turrets or X-Payment header'
```

Comment:
Otherwise use the default method of the turret list being used to insert the fee payment ops in the contract itself.